[1,2단계 - 체스] 안나(김민겸) 미션 제출합니다.

Mingyum-Kim opened the pull request:
# 👀 구현 과정

## 📝 객체 설계

### [체스 이동 관련 요소]

**ChessBoard**

- 기물이 어떤 칸에 위치하는 지 저장한다.
- 기물을 다른 칸으로 이동시킨다.

**Square**

- 체스판의 한 칸을 나타낸다.

**Movable**

- 기물이 이동할 수 있는 조건을 저장한다.
- `direction` 방향으로 `maxStepSize` 까지만 이동할 수 있다.

**Direction** 

- 기물들이 이동할 수 있는 모든 방향을 정의한다.
- `sourcePosition` 과 `targetPosition` 의 방향을 결정한다.

**Vector**

- `sourcePosition` 에서 `targetPosition` 로 이동하는 단위 벡터를 저장한다.
- 해당 단위 벡터 값을 기반으로 Direction을 결정할 수 있다.

### [체스판의 위치 관련 요소]

**Position**

- 체스판의 위치를 저장한다.

**File**

- 체스판의 Column Index를 저장한다.

**Rank**

- 체스판의 Row Index를 저장한다.

### [기물 관련 요소]

**Piece**

- 기물의 종류 (킹, 퀸 .. )과 진영(흑, 백)을 가진다.

**PieceRole**

- 6가지의 기물을 추상화한 기물 종류를 나타내는 인터페이스
- 해당 기물이 `source`에서 `target`으로 이동 가능한지 체크한다.

**PieceGenerator**

- 게임 시작 후 기물을 초기화한다.

## 🛠 애플리케이션 흐름도

### **기물을 이동하는 기능 흐름도**

![image](https://github.com/Comedy-Coding-Club/java-chess/assets/67851124/a8a42539-e79a-428a-8816-71bee9e249fb)

### 기물 이동을 위한 검증 로직

**ChessBoard 에서의 검증**

ChessBoard는 유일하게 체스판마다 어떤 기물이 있는 지 저장하고 있어요. 그래서 기물 이동을 위한 검증 로직 중 **다른 기물의 위치**를 알아야만 하는 로직은 모두 ChessBoard에서 처리하게 되었습니다.  

- 모든 기물들에 대한 검증
    - source 위치에 기물이 없으면 잘못된 기물 이동 요청이다.
    - target 위치에 같은 진영의 기물이 있으면 이동할 수 없다.
    - source 위치와 target 위치가 같으면 잘못된 입력이다.
- Pawn 기물에 대한 검증
    - 흑색인 경우 아래로 전진하고 백색인 경우 위로 전진한다.
    - 위 (혹은 아래)로 이동하려는 경우, 다른 기물이 있으면 이동할 수 없다.
    - 대각선 위 (혹은 아래)로 이동하려는 경우, 다른 기물이 있으면 이동할 수 없다.
- Knight 기물이 아닌 모든 기물들에 대한 검증
    - Knight 기물을 제외하고 모든 기물은 이동 경로에 다른 기물이 있으면 이동할 수 없다.

**PieceRole에서의 검증**

해당 기물이 **입력 받은 곳으로 움직일 수 있는지**를 검사해야 합니다. 예를 들어, 킹은 대각선으로 가거나 두 칸 이상 전진할 수 없습니다. 이런 경우 킹을 대각선으로 이동하려 하면 이동 요청을 무효화하고 킹을 이동시키지 않았습니다.

이렇게 기물마다 다른 이동 조건을 PieceRole을 구현한 각각의 기물 클래스에서 정의하였습니다. 조건을 검증하는 흐름은 아래와 같습니다. 

- 입력한 source, target Position을 기반으로 이동할 방향을 결정합니다. (`Direction`)
- 입력한 source, target Position을 기반으로 기물이 몇 칸 이동해야하는 지를 계산합니다. 이 때 기물의 **이동 벡터**가  `Direction`에 정의된 **단위 벡터**의 몇 배인지 구합니다.
- 기물 클래스에 정의된 `List<Movable>` 중 하나라도 방향이 일치하면서 `maxStepSize`보다 작거나 같은 이동 횟수라면 **이동이 가능하다**라고 판단합니다.

**이동 전략 **

- 킹 : 상하좌우 한 칸 이동 가능
- 퀸 : 상하좌우와 대각선 모든 방향으로 무한 칸 이동 가능
- 비숍 : 대각선 네 방향으로 무한 칸 이동 가능
- 나이트 : L 자 모양으로 어느 방향으로든 이동 가능
- 룩 : 상하좌우 무한 칸 이동 가능
- 폰 : 첫 움직임에서만 두 칸 이동 가능. 그 이후로는 한 칸 이동 가능.

보다 자세한 설명은 [나무 위키](https://namu.wiki/w/%EC%B2%B4%EC%8A%A4)를 참고해주세요 🥰

Mingyum-Kim committed:
docs: 기능 구현 목록 작성

Mingyum-Kim committed:
feat(Rank): 체스판의 가로줄을 생성하는 기능 구현

xogns1514 committed:
feat(ChessBoard): 체스판을 생성하는 기능 구현

xogns1514 committed:
feat(Position): 칸의 위치를 생성하는 기능 구현

Mingyum-Kim committed:
feat(Square): 체스판의 한 칸을 생성하고 동등성을 비교하는 기능 구현

xogns1514 committed:
feat(Piece): 기물을 나타내는 도메인을 생성

xogns1514 committed:
feat(PieceType): 기물의 종류를 나타내는 도메인 생성

xogns1514 committed:
feat(Piece): 기물의 종류와 색깔을 포함하는 도메인 생성

xogns1514 committed:
feat(Rank): Rank 필드값을 List<Square>로 변경

Mingyum-Kim committed:
refactor(Square): PieceType이 아닌 Piece를 필드로 가지도록 수정

Mingyum-Kim committed:
feat(RankGenerator): 체스판을 초기화하는 기능 구현

xogns1514 committed:
feat(OutputView): 체스판을 출력하는 기능 구현

Mingyum-Kim committed:
feat(ChessController): 체스를 출력하는 함수를 호출하는 컨트롤러 구현

Mingyum-Kim committed:
refactor(OutputView): 레터링과 넘버링을 제외하여 출력하도록 수정

Mingyum-Kim committed:
feat(InputView): 체스 게임을 시작하는 명령어를 받고 출력하는 기능 구현

xogns1514 committed:
style: 문자열 포맷팅

xogns1514 committed:
refactor(PieceType): PieceType 클래스명 PieceRole로 변경

xogns1514 committed:
refactor(Piece): Piece 클래스명 PieceType으로 변경

xogns1514 committed:
chore: 패키지 분리

xogns1514 committed:
feat(RankGenerator): generatePieceType 메서드 가독성 개선

Mingyum-Kim committed:
docs: 2단계 미션 기능 구현 목록 작성

Mingyum-Kim committed:
Merge branch 'step1' of https://github.com/Mingyum-Kim/java-chess into step1

Mingyum-Kim committed:
feat(Piece): 기물의 유형과 위치를 저장하는 Piece 객체 구현

xogns1514 committed:
feat(RankGenerator): generate 메서드 Piece 의존성 제거

xogns1514 committed:
feat(Square): PieceType 필드 제거

xogns1514 committed:
feat(PieceMover): piece가 source에 위치한 piece를 target으로 이동하는 기능 구현

xogns1514 committed:
feat(OutputView): 불필요한 메서드 제거

Mingyum-Kim committed:
feat(PieceGenerator): 기물을 초기화하는 기능 구현

Mingyum-Kim committed:
feat(OutputView): PieceMover의 정보를 기반으로 체스판의 상태를 출력하도록 구현

xogns1514 committed:
feat(Position): 체스판의 row와 column을 File, Rank 클래스로 래핑

Mingyum-Kim committed:
feat(PieceMover): 기물을 이동시킬 때 생길 수 있는 예외를 처리

Mingyum-Kim committed:
test(PieceMover): 에러 상황에 대한 테스트 코드 작성

xogns1514 committed:
feat(Direction): source와 target을 통해 이동할 방향을 찾는 기능 구현

xogns1514 committed:
feat(King): King의 움직임 수 있는 방향을 판단하는 기능 구현

xogns1514 committed:
fix: 에러나는 부분 주석 처리

Mingyum-Kim committed:
feat(PieceRole): Piece의 움직임 방향을 정의

xogns1514 committed:
refactor: 변경된 Piece에 따라 체스판 출력하도록 수정

Mingyum-Kim committed:
docs: 기능 구현 목록 작성

Mingyum-Kim committed:
feat(ChessController): 기물을 이동하고 체스판을 출력하는 기능 구현

xogns1514 committed:
test: 기물의 움직임 가능 여부 테스트

Mingyum-Kim committed:
feat(Pawn): 폰이 이동 가능한 경로를 조회하는 기능 구현

Mingyum-Kim committed:
refactor(Direction): Direction의 결정 조건을 올바르게 수정

xogns1514 committed:
test: 폰의 움직임 가능 여부 테스트

Mingyum-Kim committed:
feat(Pawn): Pawn가 움직일 때 첫 번째 턴에서만 두 칸 이동할 수 있도록 설정

xogns1514 committed:
feat(PieceMover): 나이트를 제외한 기물은 이동경로에 기물이 있으면 이동불가 기능 구현

xogns1514 committed:
feat(ChessController): 올바르지 않은 move 명령어 예외처리

xogns1514 committed:
refactor: 코드 가독성 개선

xogns1514 committed:
refactor(PieceMover): 기물 이동 예외 메서드 분리

xogns1514 committed:
refactor(OutputView): 체스판 출력 메서드 분리

xogns1514 committed:
refactor(PieceType): Piece의 Position을 제거하고 PieceType을 Piece로 변경

xogns1514 committed:
refactor(Movable): 기물의 최대 이동 횟수를 검증하는 메서드 분리

xogns1514 committed:
refactor(Pawn): Pawn 색깔에 따라 이동경로 분리

xogns1514 committed:
refactor(PieceMover): PieceMover 클래스명 ChessBoard로 변경

Mingyum-Kim committed:
refactor(PieceGenerator): 가로 단위의 Piece를 초기화하는 로직 개선

Mingyum-Kim committed:
refactor(Vector): 방향을 결정하는 file, rank의 차이 도메인을 Vector라고 지정

Mingyum-Kim committed:
refactor: 디렉터리 구조 변경

Mingyum-Kim committed:
refactor(ChessController): 컨트롤러 메서드의 indent depth 감소

Mingyum-Kim committed:
feat(OutputView): 체스 게임을 시작할 때 메시지를 출력

Mingyum-Kim committed:
refactor(Square): Square를 Record로 구현

Mingyum-Kim committed:
refactor(Piece): Piece 객체를 Record로 변경

Mingyum-Kim committed:
refactor(Queen)): 퀸 기물에 대해 이동 가능한 방향을 추가

Mingyum-Kim committed:
refactor: 기물이 이동할 수 없는 곳이면 예외를 발생시키도록 수정

나이트를 제외한 다른 기물들이 이동할 수 있는 지 검증하는 로직을 PieceRole로 이동

Mingyum-Kim committed:
refactor: 폰이 전진하거나 대각선으로 이동하는 경우의 예외 처리를 Pawn에서 구현

Mingyum-Kim committed:
refactor: 테스트 코드의 기물 생성 부분을 정적 팩토리 메서드로 변경

Mingyum-Kim committed:
refactor(ChessController): 체스를 실행하는 기능을 하는 메서드 이름을 변경

Mingyum-Kim committed:
refactor: 체스판의 기물을 이동할 수 없을 때 IllegalArgumentException을 반환하도록

Mingyum-Kim committed:
refactor: Square를 삭제하고 Position만으로 기물의 위치를 확인하도록 수정

Mingyum-Kim committed:
refactor: 반복적으로 패턴 객체를 생성하는 문제를 해결

Mingyum-Kim committed:
refactor(ChessController): 잘못된 입력 시 재입력을 받을 수 있도록 구현

Mingyum-Kim committed:
refactor: Turn을 추가하여 자신의 차례일 때만 이동할 수 있도록 구현

Mingyum-Kim committed:
refactor: 처음부터 `이동` 입력이 들어오면 예외를 발생시키도록 수정

Mingyum-Kim committed:
refactor: enterCommand 메서드를 더 명확하게 작성

Mingyum-Kim committed:
test: Movable 테스트 케이스 작성

Mingyum-Kim committed:
feat(DirectionTest): 방향 탐색에 대한 테스트 케이스 추가

JINU-CHANG commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,14 @@
+    private final Scanner scanner = new Scanner(System.in);
+    public static final String MOVE_POSITION_REGEX_FORMAT = "^[a-h][1-8]$";
+
+    public List<String> enterStartOrEndOrMoveCommand() {
```
메서드명에 모든 Command를 명시해줄 필요가 있을까요??

JINU-CHANG commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,93 @@
+        ChessCommand initialCommand = enterCommandAndExtractPrefix();
+
+        if (initialCommand == ChessCommand.START) {
+            ChessBoard chessBoard = new ChessBoard();
```
정적 팩토리 메서드를 통해 생성시 체스판을 초기화한다는 의미가 더 드러나면 좋을 것 같아요!!

jhon3242 commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,93 @@
+        }
+        if (initialCommand == ChessCommand.MOVE) {
+            throw new IllegalStateException("게임이 시작되지 않았습니다.");
+        }
```
👍👍

jhon3242 commented on src/main/java/domain/position/File.java:
```diff
@@ -0,0 +1,41 @@
+
+    public File add(final int movement) {
+        return new File((char) (letter + movement));
+    }
```
저라면 개인적으로 `letter + movement < 'a' || letter + movement > 'h'` 인 경우 예외를 터트리도록 처리했을 것 같은데 따로 검증을 해주지 않으신건 의도된 설계이신지 궁금합니다!

jhon3242 commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,93 @@
+
+    private boolean isNotValidCoordinateInput(final String input, final Pattern pattern) {
+        return !pattern.matcher(input).matches();
+    }
```
Pattern 대신 Matcher 를 받았으면 성능상 더 좋을 것 같아요! matcher 메서드가 호출될 때 마다 새로운 객체를 생성하기 떄문이죠! ㅎㅎ 자세한 내용은[ 커비님의 블로그](https://velog.io/@dlguswl936/%EA%B3%B5%EB%B6%80%ED%95%9C-%EA%B2%83-String.matches%EB%8A%94-%EC%99%9C-%EC%84%B1%EB%8A%A5%EC%97%90-%EC%95%88-%EC%A2%8B%EC%9D%84%EA%B9%8C)를 첨부합니다.

jhon3242 replied:
추가로 여기서 그냥 `Pattern.compile(InputView.MOVE_POSITION_REGEX_FORMAT);` 를 호출해줘도 괜찮지 않을까 싶네요! 그러면 파라미터로 매번 Pattern 을 넘겨줄 필요가 없으니까요 😁

jhon3242 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,14 @@
+    private final Scanner scanner = new Scanner(System.in);
+    public static final String MOVE_POSITION_REGEX_FORMAT = "^[a-h][1-8]$";
+
+    public List<String> enterStartOrEndOrMoveCommand() {
+        String input = scanner.nextLine();
+        return List.of(input.split(" "));
+    }
```
지극히 개인적인 의견이지만 enter 이라는 Prefix 는 "들어간다" 라는 느낌이 드는 것 같아요!
문자열을 읽어 들인다는 의미로 input 이나 read 같은 Prefix 는 어떨까요?? 
(지극히 개인적인 의견이니까 별로면 무시해도 좋아요 근데 내 마음의 상처는 받을듯)

jhon3242 commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private Position generateSquare(final String coordinate, final Pattern pattern) {
+        if (isNotValidCoordinateInput(coordinate, pattern)) {
+            throw new IllegalArgumentException("이동할 source와 target 정보를 다시 입력해주세요.");
+        }
+
```
이 작업은 이전에 호출된 `validateMoveCommand` 에서 해주면 어떨까요?? 
이미 앞서 검증 메서드가 호출되었는데 또 검증을 하는 것이 어색하다고 느껴집니다!

jhon3242 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private void commonMoveValidate(Position sourcePosition, Position targetPosition) {
+        if (!piecePosition.containsKey(sourcePosition)) {
+            throw new IllegalArgumentException("해당 위치에 Piece가 존재하지 않습니다.");
+        }
+        if (hasSameColorPiece(sourcePosition, targetPosition)) {
+            throw new IllegalArgumentException("같은 진영의 기물이 있는 곳으로 옮길 수 없습니다.");
+        }
+        if (sourcePosition.equals(targetPosition)) {
+            throw new IllegalArgumentException("같은 위치로의 이동입니다. 다시 입력해주세요.");
+        }
```
상황에 따른 예외 메시지를 구분해주니까 예외 메시지만 봐도 어떤 문제인지 바로 파악할 수 있겠네요! 👍

jhon3242 commented on src/main/java/domain/piece/piecerole/PieceRole.java:
```diff
@@ -0,0 +1,62 @@
+
+    protected void validateCorrectRouteForPiece(final Position source,
+                                                final Position target) {
+        boolean cannotMove = routes.stream()
+                .noneMatch(movable -> movable.canMove(source, target));
```
noneMatch 배워갑니다! 👍

jhon3242 commented on src/main/java/domain/position/Position.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public Position move(final Direction direction) {
+        int dFile = direction.getFileVector();
+        int dRank = direction.getRankVector();
```
삐빅 축약입니다🚨

jhon3242 commented on src/main/java/domain/game/Direction.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+
+    public static Direction findDirection(Position source, Position target) {
+        Vector vector = target.subtract(source);
+        return Arrays.stream(values())
+                .filter(direction -> isSameDirection(direction, vector))
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("움직일 수 없는 방향입니다."));
+    }
```
👍👍 이 팀에는 이과가 있는게 분명하네요. 벡터를 활용한 부분이 인상깊습니다!

jhon3242 commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private void update(Position sourcePosition, Position targetPosition, Piece findPiece) {
+        if (piecePosition.containsKey(targetPosition)) {
+            piecePosition.remove(targetPosition);
+        }
```
이 부분은 없어도 될 것 같아요! 어차피 밑에서 put 으로 업데이트 해주니까요😎

JINU-CHANG replied:
이렇게 줄여볼 수 있을 것 같습니다 !!
```java
piecePosition.remove(sourcePosition);
piecePosition.put(targetPosition, findPiece);
```

jhon3242 reviewed (commented):
안녕하세요 안나! 타칸입니다!
코드 잘 읽었습니다! 
벡터를 활용한 부분이 인상깊었습니다!
스텝 3, 4도 기대하겠습니다! 😋

아 그리고 브랜치를 전부 다 가져오셨더라고요 ㅎㅎ 
git checkout 했는데 브랜치가 389개가 나와서 당황스러웠답니다! 🙃

JINU-CHANG commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,93 @@
+            List<String> input = inputView.enterStartOrEndOrMoveCommand();
+            command = ChessCommand.from(input.get(0));
+
+            executeMoveCommand(chessBoard, input, command, pattern);
```
움직임을 시행하는 메서드인데 input, pattern이 파라미터로 넘겨지는게 한 번에 이해되지 않는 것 같아요 🤔

JINU-CHANG commented on src/main/java/controller/ChessController.java:
```diff
@@ -0,0 +1,93 @@
+import view.InputView;
+import view.OutputView;
+
+public class ChessController {
```
컨트롤러 로직 꼼꼼하게 작성하신 게 느껴집니다 👍👍

JINU-CHANG commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,77 @@
+        PieceGenerator.generate(piecePosition);
+    }
+
+    public void add(Position position, Piece piece) {
```
final 붙인 곳과 안 붙인 곳 통일시켜주세요-!-!

JINU-CHANG commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,77 @@
+        commonMoveValidate(sourcePosition, targetPosition);
+
+        Piece findPiece = piecePosition.get(sourcePosition);
+        validateCorrectTurn(findPiece);
```
👍👍

JINU-CHANG replied:
![image](https://github.com/Comedy-Coding-Club/java-chess/assets/98975580/3ecbb365-f1fa-4740-8343-cc615cb18e50)

쓱 봤을 때 에러 메시지인지 한번에 이해하지 못했는데요, 에러 메시지임을 좀 더 강조하면 좋을 것 같아요 🙂

[ERROR] 를 붙이거나, Black을 움직일 수 없습니다. 라고 해보는 건 어떨까요 ??

JINU-CHANG commented on src/main/java/domain/game/ChessBoard.java:
```diff
@@ -0,0 +1,77 @@
+
+public class ChessBoard {
+    private Color turn;
+    private final Map<Position, Piece> piecePosition;
```
piecePositions라고 복수형으로 이름 붙이지 않은 이유가 있나요~?? 코드를 읽다가 `piecePosition`이 한 개의 기물 위치를 나타내는 줄 알았는데 그게 아니여서 헷갈리더라구요

JINU-CHANG commented on src/main/java/domain/piece/piecerole/PieceRole.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    protected void validateCorrectRouteForPiece(final Position source,
```
piece가 `List<Movable>`을 가지고 있음으로써 로직이 정말 깔끔해진 것 같아요 👏

JINU-CHANG reviewed (commented):
안나~~~ 제제입니다 😊
친절한 리드미와 깔끔한 코드 덕분에 처음보는 코드임에도 이해하기 수월했던 것 같아요 👍
복잡한 로직들을 잘 풀어주신 것 같아서 배울 부분이 많았습니다!
사소한 부분에서 궁금했던 것들 피드백 남겼어요 시간나실 때 천천히 확인해주세요~

Mingyum-Kim committed:
refactor: Vector에서 unit vector를 계산하여 반환하도록 수정

Mingyum-Kim committed:
refactor(ChessController): 뷰 객체의 생성을 Controller에서 구현

Mingyum-Kim committed:
refactor: 컨트롤러 내부에서 뷰 객체를 생성하도록 수정

Mingyum-Kim committed:
refactor(SeparatedInput): 뷰에서 start, end, move 입력을 하나의 뷰 클래스에서 받아오도록 수정

Mingyum-Kim committed:
refactor: 명령어에 따라 서로 다른 동작을 하는 Executor를 정의하고 컨트롤러 로직 개선

Mingyum-Kim committed:
refactor: 에러 메시지의 prefix에 `[ERROR]` 추가

Mingyum-Kim committed:
refactor: Direction 클래스의 주석 제거

Mingyum-Kim committed:
refactor: 게임 진행 중 에러가 발생하면 에러를 출력하고 종료하도록 수정

StackTrace를 출력하지 않기 위해 Controller에서 try catch문을 작성한다.

Mingyum-Kim committed:
refactor(ChessBoardGenerator): 체스판을 초기화하는 Generator가 ChessBoard를 반환하도록 수정

Mingyum-Kim committed:
refactor(PositionFixture): Fixture를 반환하는 메소드의 네이밍을 간략하게 수정

Mingyum-Kim committed:
refactor: Movable과 Pawn을 VO로 변경하여 기물의 상태가 변하지 않도록 구현

Pawn을 VO로 구현하기 위해서 변경될 수있는 부분인 List<Movable>도 VO로 구현하였습니다. 이 과정에서 WhitePawn, BlackPawn마다 가지고 있는 Movable이 다르므로 상속하여 따로 구현하였습니다.

Mingyum-Kim committed:
feat(Turn): ChessBoard가 가지고 있는 Turn을 래퍼 클래스로 구현

Mingyum-Kim committed:
refactor(ChessBoardTest): findPiece를 일괄 Piece로 수정

Mingyum-Kim committed:
refactor: Assertions를 static import하여 수정

Mingyum-Kim committed:
refactor: 상수의 네이밍을 대문자로 수정

Mingyum-Kim committed:
feat(PieceMapper): 뷰의 출력을 PieceMapper가 담당하도록 수정

Mingyum-Kim committed:
refactor(ChessBoard): 이동하는 위치에 다른 진영의 기물이 있는 경우 제거하는 로직 삭제

Mingyum-Kim committed:
refactor: 모든 매개변수에 일괄적으로 final 키워드를 추가

Mingyum-Kim committed:
refactor: sourcePosition, targetPosition 이라는 파라미터 명을 source, target으로 수정

Mingyum-Kim committed:
refactor: File, Rank, Movable 도메인 생성자에서 입력값을 검증

Mingyum-Kim committed:
refactor(Pawn): WhitePawn과 BlackPawn의 중복 코드를 Pawn으로 이동

Mingyum-Kim committed:
refactor(SeparatedCommandInput): 명령어 입력을 쪼개서 저장하는 클래스 명을 변경

Mingyum-Kim committed:
refactor: 기물을 생성하는 정적 팩토리 메서드 네이밍을 create()로 수정

Mingyum-Kim committed:
refactor: Enum 클래스의 마지막 상수에 쉼표를 붙이고, 한 줄 개행하여 세미콜론 붙이기

Mingyum-Kim committed:
refactor: 코드 내의 매직 넘버를 상수화하여 처리

Mingyum-Kim committed:
refactor: 코드 내의 매직 넘버를 상수화하여 처리

Mingyum-Kim committed:
refactor: 열 제한을 초과하는 코드의 개행 스타일 통일

Mingyum-Kim committed:
Merge branch 'step1' of https://github.com/Mingyum-Kim/java-chess into step1

Mingyum-Kim committed:
fix: 테스트 코드에서 에러가 생기는 부분을 수정

Mingyum-Kim committed:
test: 테스트 코드 추가
